pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BrokerRequestHandler.java

```diff
@@ -36,4 +39,19 @@ public interface BrokerRequestHandler {
   BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentity requesterIdentity,
       RequestContext requestContext)
       throws Exception;
+
+  Map<Long, String> getRunningQueries();
+
+  /**
+   * Cancel a query as identified by the queryId. This method is non-blocking so the query may still run for a while
+   * after calling this method. This cancel method can be called multiple times.
+   * @param queryId the unique Id assigned to the query by the broker
+   * @param timeoutMs timeout to wait for servers to respond the cancel requests
+   * @param executor to send cancel requests to servers in parallel
+   * @param connMgr to provide the http connections
+   * @param serverResponses to collect cancel responses from all servers if a map is provided
+   * @return true if there is a running query for the given queryId.
+   */
+  boolean cancelQuery(long queryId, int timeoutMs, Executor executor, HttpConnectionManager connMgr,
```

Comment:
Should it throw `TimeoutException` (and maybe `InterruptedException`) when the server didn't return within the timeout?

Reply:
good point. there is Exception but logged out as you commented below. Instead, I'll throw it out.